src/core/server/ui_settings/saved_objects/permission_controlled_ui_settings_wrapper.ts

```diff
@@ -0,0 +1,172 @@
+      );
+
+      if (adminSettingObject) {
+        throw new Error('Bulk create is not supported for admin settings');
+      }
+
+      return wrapperOptions.client.bulkCreate<T>(objects, options);
+    };
+
+    const bulkUpdateWithPermission = async <T = unknown>(
+      objects: Array<SavedObjectsBulkUpdateObject<T>>,
+      options?: SavedObjectsBulkUpdateOptions
+    ) => {
+      const adminSettingObject = objects.find(
+        (obj) => obj.type === 'config' && obj.id === DASHBOARD_ADMIN_SETTINGS_ID
+      );
+
+      if (adminSettingObject) {
+        throw new Error('Bulk update is not supported for admin settings');
+      }
+
+      return wrapperOptions.client.bulkUpdate<T>(objects, options);
+    };
+
+    return {
+      ...wrapperOptions.client,
+      create: createUiSettingWithPermission,
+      bulkCreate: bulkCreateWithPermission,
+      delete: wrapperOptions.client.delete,
+      update: updateUiSettingsWithPermission,
+      bulkUpdate: bulkUpdateWithPermission,
+      get: wrapperOptions.client.get,
+      checkConflicts: wrapperOptions.client.checkConflicts,
+      errors: wrapperOptions.client.errors,
+      addToNamespaces: wrapperOptions.client.addToNamespaces,
+      deleteFromNamespaces: wrapperOptions.client.deleteFromNamespaces,
+      find: wrapperOptions.client.find,
+      bulkGet: wrapperOptions.client.bulkGet,
+      deleteByWorkspace: wrapperOptions.client.deleteByWorkspace,
+    };
+  };
+
+  private checkAdminPermission(wrapperOptions: SavedObjectsClientWrapperOptions): void {
+    // If saved object permission is disabled, everyone should be treated as admin here
+    const isDashboardAdmin =
+      getWorkspaceState(wrapperOptions.request).isDashboardAdmin !== false ||
+      !this.isPermissionControlEnabled;
+
+    if (!isDashboardAdmin) {
+      throw SavedObjectsErrorHelpers.decorateForbiddenError(
+        new Error('No permission for admin UI settings operations')
+      );
+    }
+  }
+
+  private getAclInstance(): ACL | undefined {
+    if (!this.isPermissionControlEnabled) {
+      return undefined;
+    }
+
+    if (!this.aclInstance) {
+      // Allow read for all users but no write for any users
+      // This means only dashboard admin would have permission to bypass ACL
+      this.aclInstance = new ACL().addPermission(['read'], {
+        users: ['*'],
+      });
+    }
+
+    return this.aclInstance;
+  }
+
+  private async createPermissionUiSetting<T>(
+    attributes: T,
+    options: SavedObjectsBaseOptions,
+    wrapperOptions: SavedObjectsClientWrapperOptions
+  ): Promise<SavedObject<T>> {
+    this.checkAdminPermission(wrapperOptions);
+
+    const aclInstance = this.getAclInstance();
+
+    return wrapperOptions.client.create('config', attributes, {
+      ...options,
+      overwrite: true,
+      id: DASHBOARD_ADMIN_SETTINGS_ID,
+      ...(aclInstance ? { permissions: aclInstance.getPermissions() } : {}),
```

Comment:
Did you consider the case that the admin ui setting was initialized with permission disabled, but later object permission is enabled?

Reply:
Good question. In this case, we will still disable the admin UI setting on the advanced setting page (still we also do the saved permission checking on frontend) but the server side will skip the ACL permission check. I think we can leave it in this PR and properly address this problem in a later PR since this one already being quite big.